File: src/services/docs.ts
import { invoke } from "@tauri-apps/api/core";
import type { Doc } from "@/models/doc";

export async function loadDocs(vaultPath: string) {
  if (!vaultPath) {
    return [] as Doc[];
  }

  return invoke<Doc[]>("list_docs", {
    path: vaultPath,
  });
}

export async function createDoc(vaultPath: string, doc: Doc) {
  if (!vaultPath) {
    return doc;
  }

  return invoke<Doc>("create_doc", {
    path: vaultPath,
    doc,
  });
}

export async function updateDoc(vaultPath: string, doc: Doc) {
  if (!vaultPath) {
    return doc;
  }

  return invoke<Doc>("update_doc", {
    path: vaultPath,
    doc,
  });
}

export async function deleteDoc(vaultPath: string, id: string) {
  if (!vaultPath) {
    return;
  }

  await invoke("delete_doc", {
    path: vaultPath,
    id,
  });
}

File: src-tauri/src/commands/docs.rs
use crate::application::vault::open_database_from_input;
use crate::persistence::doc_repository::DocRepository;
use crate::persistence::models::Doc;

#[tauri::command]
pub fn list_docs(path: &str) -> Result<Vec<Doc>, String> {
    let db = open_database_from_input(path)?;
    DocRepository::new(&db).list()
}

#[tauri::command]
pub fn create_doc(path: &str, doc: Doc) -> Result<Doc, String> {
    let db = open_database_from_input(path)?;
    DocRepository::new(&db).create(doc.clone())?;
    Ok(doc)
}

#[tauri::command]
pub fn update_doc(path: &str, doc: Doc) -> Result<Doc, String> {
    let db = open_database_from_input(path)?;
    DocRepository::new(&db).update(doc.clone())?;
    Ok(doc)
}

#[tauri::command]
pub fn delete_doc(path: &str, id: &str) -> Result<(), String> {
    let db = open_database_from_input(path)?;
    DocRepository::new(&db).delete(id)
}
